src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -5,45 +5,126 @@
-            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
-
-            compiler.evaluate(ast)
-            compiler.compile_additions
-            compiler.catalog_data_hash
+      def self.compile_ast(code, compile_options, boltlib_path)
+        # If the request requires that bolt be loaded we assume we are in a properly 
+        # configured PE environment
+        if compile_options.dig('options', 'compile_for_plan')
+          # CODEREVIEW: Where to we handle this error and what should it say?
+          unless boltlib_path
+            msg = "The path to boltlib modules must be provided in jruby-puppet boltlib-path puppetserver setting"
+            raise Puppet::Error, _(msg)
+          end
+
+          require 'bolt/apply_inventory'
+          require 'bolt/apply_target'
+
+          Puppet[:node_name_value] = compile_options['certname']
+
+          # Use the existing environment with the requested name
+          Puppet::Pal.in_environment(compile_options['environment'],
+                                     pre_modulepath: boltlib_path,
+                                     envpath: Puppet[:environmentpath],
+                                     facts: compile_options['facts']['values']) do |pal|
+            # TODO: Given we hide this from plan authors this current iteration has only
+            # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
+            # merged and promoted we can just use an empty hash.
+            fake_config = {
+              'transport' => 'redacted',
+              'transports' => {
+                'redacted' => 'redacted'
+              }
+            }
+            bolt_inv = Bolt::ApplyInventory.new(fake_config)
+            Puppet.override(bolt_inventory: bolt_inv) do
+              Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+              # This compiler has been configured with a node containing
+              # the requested environment, facts, and variables, and is used
+              # to compile a catalog in that context from the supplied AST.
+              pal.with_catalog_compiler do |compiler|
+                # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+                Puppet[:strict] = :warning
+                Puppet[:strict_variables] = false
+                vars = Puppet::Pops::Serialization::FromDataConverter.convert(compile_options['variables']['values'])
+
+                # Facts trump plan variables and target vars.
+                # See https://github.com/puppetlabs/bolt/issues/1725 for details
+                compile_options['facts']['values'].keys.each {|fact_name| vars.delete(fact_name)}
+                # TODO: PUP-10476 Refactor PAL api such that we do not need to use private methods
+                pal.send(:add_variables, compiler.send(:topscope), vars)
```

Comment:
Promising potential solution for this https://github.com/puppetlabs/bolt/pull/1784